contrib/config/test/kv_store_xds_delegate_integration_test.cc

```diff
@@ -0,0 +1,340 @@
+    auto entries = loader->getObject("entries");
+    if (entries->hasObject(key)) {
+      return entries->getObject(key)->getString("final_value");
+    }
+    return "";
+  }
+
+  void checkSecretExists(const std::string& secret_name, const std::string& version_info) {
+    auto response = IntegrationUtil::makeSingleRequest(
+        lookupPort("admin"), "GET", "/config_dump?resource=dynamic_active_secrets", "",
+        downstreamProtocol(), version_);
+    EXPECT_TRUE(response->complete());
+    EXPECT_EQ("200", response->headers().getStatusValue());
+    Json::ObjectSharedPtr loader = TestEnvironment::jsonLoadFromString(response->body());
+    envoy::admin::v3::ConfigDump config_dump;
+    TestUtility::loadFromJson(loader->asJsonString(), config_dump);
+    // Expect at least the "client_cert" dynamic secret.
+    ASSERT_GE(config_dump.configs_size(), 1);
+    envoy::admin::v3::SecretsConfigDump::DynamicSecret dynamic_secret;
+    ASSERT_OK(MessageUtil::unpackToNoThrow(config_dump.configs(0), dynamic_secret));
+    EXPECT_EQ(secret_name, dynamic_secret.name());
+    EXPECT_EQ(version_info, dynamic_secret.version_info());
+  }
+
+  void shutdownAndRestartTestServer() {
+    // Reset the test server.
+    test_server_.reset();
+    on_server_init_function_ = nullptr;
+
+    // Set up a new Envoy, using the previous Envoy's configuration, and create the test server.
+    ConfigHelper helper(version_, *api_,
+                        MessageUtil::getJsonStringFromMessageOrDie(config_helper_.bootstrap()));
+    std::vector<uint32_t> ports;
+    std::vector<uint32_t> zero;
+    for (auto& upstream : fake_upstreams_) {
+      if (upstream->localAddress()->ip()) {
+        ports.push_back(upstream->localAddress()->ip()->port());
+        zero.push_back(0);
+      }
+    }
+    helper.setPorts(zero, true); // Zero out ports set by config_helper_'s finalize();
+    const std::string bootstrap_path = finalizeConfigWithPorts(helper, ports, use_lds_);
+
+    std::vector<std::string> named_ports;
+    const auto& static_resources = config_helper_.bootstrap().static_resources();
+    named_ports.reserve(static_resources.listeners_size());
+    for (int i = 0; i < static_resources.listeners_size(); ++i) {
+      named_ports.push_back(static_resources.listeners(i).name());
+    }
+
+    // Simulate the upstream xDS servers going down.
+    closeConnection(sds_connection_);
+    closeConnection(rtds_connection_);
+    getSdsUpstream().reset();
+    getRtdsUpstream().reset();
+
+    // Create and start the new Envoy.
+    createGeneratedApiTestServer(bootstrap_path, named_ports, {false, true, false}, false,
+                                 test_server_);
+    registerTestServerPorts(named_ports, test_server_);
+  }
+
+  FakeHttpConnectionPtr sds_connection_;
+  FakeStreamPtr sds_stream_;
+  FakeHttpConnectionPtr rtds_connection_;
+  FakeStreamPtr rtds_stream_;
+};
+
+INSTANTIATE_TEST_SUITE_P(IpVersions, KeyValueStoreXdsDelegateIntegrationTest,
```

Comment:
This should also be tested against unified-SotW

Reply:
This PR doesn't handle unified-SotW yet.  I wanted to get it working for regular Sotw and then do the same for unified-Sotw in a follow up PR.